### ISSUE TITLE
wined3d-TextureAllocation patch (contiguous mipchain allocation)

### ISSUE THREAD

**Issue by c10ud** [wine-compholio/wine-staging#290]
https://github.com/c10ud/wine/compare/TextureAllocation

This branch implements the minimal required modifications for resources to be allocated all at once in textures. 
A new structure, wined3d_subresource is introduced for surfaces etc., and it does not manage its memory.
d3d* code has been adjusted.
Some codepaths are still hackish or flaky but overall it kind of works (needs some advice/look from more experienced d3d developers).

tested with: d3d8 tests, d3d9 tests, league of legends (d3d9 mode)

~~~
This patch moves texture memory allocation from surfaces (subresources) to the texture itself (texture->resource, previously unused) and provides said subresources their own heap address.

The required size for all levels is calculated with a newly added utility function at texture_init.

A new flag, resource->managed controls whether the resource needs to free its own heap or not.

the upstream bug# is 34480 ( https://bugs.winehq.org/show_bug.cgi?id=34480 )

Patch tested with:
- League of Legends (d3d9 mode)
- World of Warcraft (d3d9 mode)

A few comments:
- This patch does not require CSMT however I diff'd against the full-patched wine tree in order to retain compatibility with said patchset, given its complexity

- WARNING: This functionality in pair with CSMT adds a little memory overhead since managed resources are not really freed until the texture itself is gone.

- Please note that League of Legends is currently unplayable with CSMT enabled due to soft-freezes randomly happening through the game, however this is unrelated to this patch
~~~

**Comment by titan73** [comment]
Ok. I commented on you commit.
What are the results of your tests. On which windows version does they pass? Are there some tests failing?
Stefan commented that the changes may be shaky with an example about discard maps. Did you take a look? Is this comment about csmt or non csmt version.
Apart from that that seems correct. We need to investigate the corner cases Stefan mentioned. And we need to test this. Can you send a version that fixes the issues in the commit comments so we can test it?

**Comment by slackner** [comment]
I also mentioned the following issues on IRC, here just again for completeness:

* We would really like to have a patch, which also applies on upstream Wine. The CSMT changes are huge, and all the "#ifdef" changes are completely autogenerated. There should be no need to add them manually when you split your changes into a "non-CSMT" part and a "CSMT" addition. As a first step it is completely sufficient when you concentrate on the non-CSMT part, as soon as this part is "solved" we can check again which additional modifications are required for CSMT.

* As mentioned in @titan73's comments, you should use `char *`, `BYTE *`, or maybe even `void *` instead of `ULONG_PTR`. We have really pointers here, so you should also use pointer datatypes.

* We will really need tests for all the corner cases. Testing with various games and apps is nice of course, but not sufficient for such a "critical" patch. These code paths are used by basically every game, if we change it we have to be sure that it is still correct.

**Comment by c10ud** [comment]
Hello,
unfortunately I do not have windows nor I know how to test in such environment, my tests are limited to my setup: wine-git 32bit with the entire -staging patchset on ubuntu 14.04 x64.

I only tested league of legends and world of warcraft (not affected by the original issue, I wanted to see if I broke anything).

In order to patch vanilla wine, the ifdef'd parts can be removed and that's it. I'm not sure on how to handle this patching thing so I guess I'll need some advice.

I will fix your remarks regarding types, whitespaces, etc.

As for testing: see this posting as a call for testing, I dont have any game except the ones I mentioned and tested.

**Comment by c10ud** [comment]
Here's a CSMT-less patch (v4):
https://github.com/c10ud/wine/commit/c78975fe583db085259e52beaadd0ca6eaa75d47

The test is unchanged, fixed cosmetics and some remarks, removed CSMT-specific ifdefs.
Not compile-checked as I need some sleep :)

Let me know how we do proceed with this, and test,test,test!

**Comment by titan73** [comment]
There are test VMs on which you can run the tests https://newtestbot.winehq.org/index.pl.

**Comment by titan73** [comment]
I read in one comment that Stefan initially wrote the test. That means he probably has passed the test on several windows.

**Comment by titan73** [comment]
I guess Stefan was referring to WINED3D_MAP_DISCARD flag. I just glanced at it but I hardly do the link with the allocation.

**Comment by titan73** [comment]
Last non CSMT patch apply and compile Fine. Test passes on wine.
I tested successfully some games and demos for a long time to make sure nothing happens (memory leaks for example):
- Skydrift demo
- Chernobyl demo
- Dragon Age origins
I continue testing...

**Comment by titan73** [comment]
Well I haven't encounter any regression yet. I will test some other games but I think this patch can go in 1.7.35. At least for the non csmt. What do you think?

**Comment by titan73** [comment]
Bela tested 100+ games w/o seeing regression for non csmt.
Any news on csmt?

**Comment by slackner** [comment]
I already looked at the patch and didn't notice any issues so far. However, before finally merging it I will take another look, to be sure that I didn't miss anything. What Bela probably has not tested for example is leaking memory.

In order to provide the same functionality for CSMT we will need a second patch or even multiple patches, which can be inserted/merged into the CSMT patchset at the right place. If for example patch 50, 100 and 150 mess around with resource memory allocation, and we will need the changes splitted into three parts, so that we can still compile and test each version inbetween.

@c10ud: Please note that you do not have to care about doing all the merges/rebases yourself (except you have a lot of spare time...). We can probably do that easier, I already wrote special scripts some time ago, just for a single purpose: Rebasing CSMT! ;)

**Comment by titan73** [comment]
I tested games for a long time 2/3 hours to have a chance to trigger an oom but this didn't happen. Is there a better way for that? something like a mem accounting debug channel?
For csmt c10ud told about a leaking issue iirc.

**Comment by c10ud** [comment]
These are the changes required for CSMT support:
```
diff --git a/dlls/wined3d/texture.c b/dlls/wined3d/texture.c
index 2e59657..b882650 100644
--- a/dlls/wined3d/texture.c
+++ b/dlls/wined3d/texture.c
@@ -1215,6 +1215,9 @@ static HRESULT cubetexture_init(struct wined3d_texture *texture, const struct wi
                 return hr;
             }
 
+#if defined(STAGING_CSMT)
+            surface->resource.map_heap_memory = (void *) current_heap;
+#endif /* STAGING_CSMT */
             surface->resource.heap_memory = current_heap;
             current_heap += wined3d_format_calculate_size(texture->resource.format,
                 device->surface_alignment, surface_desc.width, surface_desc.height, 1);
@@ -1382,6 +1385,9 @@ static HRESULT texture_init(struct wined3d_texture *texture, const struct wined3
             return hr;
         }
 
+#if defined(STAGING_CSMT)
+        surface->resource.map_heap_memory = (void *) current_heap;
+#endif /* STAGING_CSMT */
         surface->resource.heap_memory = current_heap;
         current_heap += wined3d_format_calculate_size(texture->resource.format,
             device->surface_alignment, surface_desc.width, surface_desc.height, 1);
@@ -1631,6 +1637,9 @@ static HRESULT volumetexture_init(struct wined3d_texture *texture, const struct
             return hr;
         }
 
+#if defined(STAGING_CSMT)
+        volume->resource.map_heap_memory = (void *) current_heap;
+#endif /* STAGING_CSMT */
         volume->resource.heap_memory = current_heap;
         current_heap += wined3d_format_calculate_size(texture->resource.format,
             device->surface_alignment, volume_desc.width, volume_desc.height, volume_desc.depth);
```

Now touching the CSMT patchset is to me like opening a can of worms, so if any of you could take the time of adding this small change, that would be great.

As for the warning about memory: from what I got, discarding surfaces (sub_resources) from sysmem without my patch leads to the allocated memory being freed asap (unless some other guard is triggered, but anyway..).

With this patch, since the block is allocated once for _all_ sub_resources, if one is discarded we cannot tear down that memory because others can be (are) still in use.

However, with this patch, when a new block of memory is allocated for a given sub_resource tied to that very same texture it will be freed by the "old" method in case a new discard or something else happens (since resource->managed will be false) so there's no leak.

The only minor (?) issue is in the fact that until the texture lives the "big" block (single memory block with all levels) will not be freed.

I even thought of handling this particular case but not knowing that well d3d or wined3d I don't know what to expect.

Does/Can discard happen for a single surface of a texture leaving all the other ones still in place? Or discard is texture-level so all sub_resources are tied to that and freed concurrently? (even without my patch).
Something like this could help speeding up a bit the process in case all levels are being discarded but the texture lives on, but I need some advice from someone with more experience with this whole thing.
```
    struct wined3d_texture *texture = surface->container;
    unsigned int i;

    if (!texture->resource.heap_memory)
        return;

    for (i = 0; i < texture->level_count; ++i) {
        if (texture->sub_resources[i]->managed)
            return;
    }
    TRACE("All sub_resources are not managed anymore, cleaning up main texture\n");
    resource_cleanup(&texture->resource);
```

Greets

**Comment by c10ud** [comment]
MSDN says:

**Comment by c10ud** [comment]
MSDN link: http://msdn.microsoft.com/en-us/library/windows/desktop/bb147263%28v=vs.85%29.aspx

> For mipmaps, D3DLOCK_DISCARD is allowed only on the top level. All levels are discarded by locking just the top level.
> This behavior is the same for volumes and cubes. For cubes, the top level and face 0 are locked.

So we could theoretically cleanup the "big" resource once top-level gets a new sub-resource allocation, judging from this

**Comment by slackner** [comment]
Before going into too much details with the CSMT-specific version I really would like to get all issues in non-CSMT resolved first.

I looked through the patches and noticed a couple of non-critical style issues, which can be easily fixed.

A more critical issue is that this patch can break video mem accounting.

Lets suppose we have the following sequence of calls:
```
resource_init(...);

**Comment by c10ud** [comment]
slackner, it looks like your assumption was right.
```
diff --git a/dlls/wined3d/resource.c b/dlls/wined3d/resource.c
index 20e2c37..841bb13 100644
--- a/dlls/wined3d/resource.c
+++ b/dlls/wined3d/resource.c
@@ -166,8 +166,11 @@ void resource_cleanup(struct wined3d_resource *resource)
 
     if (resource->pool == WINED3D_POOL_DEFAULT && d3d->flags & WINED3D_VIDMEM_ACCOUNTING && !resource->managed)
     {
-        TRACE("Decrementing device memory pool by %u.\n", resource->size);
-        adapter_adjust_memory(resource->device->adapter, (INT64)0 - resource->size);
+        if (resource->heap_memory)
+        {
+            TRACE("Decrementing device memory pool by %u.\n", resource->size);
+            adapter_adjust_memory(resource->device->adapter, (INT64)0 - resource->size);
+        }
     }
 
     wined3d_resource_free_sysmem(resource);
```
this stabilizes for me memory accounting in world of warcraft login screen and game
(looking at adapter_adjust_memory in directx.c)

**Comment by c10ud** [comment]
try 2 after IRC thoughts:
```diff --git a/dlls/wined3d/resource.c b/dlls/wined3d/resource.c
index 771fee6..c59c313 100644
--- a/dlls/wined3d/resource.c
+++ b/dlls/wined3d/resource.c
@@ -199,10 +199,16 @@ void resource_cleanup(struct wined3d_resource *resource)
 
     if (resource->pool == WINED3D_POOL_DEFAULT && d3d->flags & WINED3D_VIDMEM_ACCOUNTING && !resource->managed)
     {
-        if (resource->heap_memory)
+        switch (resource->type)
         {
-            TRACE("Decrementing device memory pool by %u.\n", resource->size);
-            adapter_adjust_memory(resource->device->adapter, (INT64)0 - resource->size);
+            case WINED3D_RTYPE_SURFACE:
+            case WINED3D_RTYPE_VOLUME:
+                TRACE("Resource was managed, not modifying adapter memory\n");
+                break;
+            default:
+                TRACE("Decrementing device memory pool by %u.\n", resource->size);
+                adapter_adjust_memory(resource->device->adapter, (INT64)0 - resource->size);
+                break;
         }
     }
 ```

**Comment by c10ud** [comment]
also the !resource->managed check can be removed in this

**Comment by c10ud** [comment]
v5, includes all remarks/findings so far (against vanilla wine): https://github.com/c10ud/wine/commit/478f9c65078797a6086053136514fc240f0f110d

**Comment by c10ud** [comment]
Experimental code that frees the big allocation once wined3d is done with it (alloc'd somewhere else):
```
diff --git a/dlls/wined3d/surface.c b/dlls/wined3d/surface.c
index 4249322..4cb0544 100644
--- a/dlls/wined3d/surface.c
+++ b/dlls/wined3d/surface.c
@@ -586,6 +586,8 @@ void surface_prepare_map_memory(struct wined3d_surface *surface)
 
 static void surface_evict_sysmem(struct wined3d_surface *surface)
 {
+    struct wined3d_texture *texture = surface->container;
+    unsigned int i;
     /* In some conditions the surface memory must not be freed:
      * WINED3D_TEXTURE_CONVERTED: Converting the data back would take too long
      * WINED3D_TEXTURE_DYNAMIC_MAP: Avoid freeing the data for performance
@@ -597,6 +599,18 @@ static void surface_evict_sysmem(struct wined3d_surface *surface)
 
     wined3d_resource_free_sysmem(&surface->resource);
     surface_invalidate_location(surface, WINED3D_LOCATION_SYSMEM);
+    if (!texture->resource.heap_memory)
+        return;
+
+    for (i = 0; i < texture->level_count; ++i)
+    {
+        if (texture->sub_resources[i]->managed)
+            return;
+    }
+    TRACE("All sub_resources are not managed anymore, cleaning up main texture: %p\n", texture);
+    wined3d_resource_free_sysmem(&texture->resource);
 }
 
 static void surface_release_client_storage(struct wined3d_surface *surface)
diff --git a/dlls/wined3d/volume.c b/dlls/wined3d/volume.c
index f2dd9e9..03f6b80 100644
--- a/dlls/wined3d/volume.c
+++ b/dlls/wined3d/volume.c
@@ -171,8 +171,22 @@ static void wined3d_volume_download_data(struct wined3d_volume *volume,
 
 static void wined3d_volume_evict_sysmem(struct wined3d_volume *volume)
 {
+    struct wined3d_texture *texture = volume->container;
+    unsigned int i;
+
     wined3d_resource_free_sysmem(&volume->resource);
     wined3d_volume_invalidate_location(volume, WINED3D_LOCATION_SYSMEM);
+
+    if (!texture->resource.heap_memory)
+        return;
+
+    for (i = 0; i < texture->level_count; ++i)
+    {
+        if (texture->sub_resources[i]->managed)
+            return;
+    }
+    TRACE("All sub_resources are not managed anymore, cleaning up main texture: %p\n", texture);
+    wined3d_resource_free_sysmem(&texture->resource);
 }
 
 static DWORD volume_access_from_location(DWORD location)
```
However I do not see system memory decreasing even if the path is reached. Weird.

**Comment by titan73** [comment]
I'm investigating the crash in wine tests. I've already found an issue where surface can be resized (typically render target and depth stencil when screen resolution changes). I fixed the issue but the crash is still there. This may take some time to figure out as it requires understanding wined3d code a bit more.

**Comment by c10ud** [comment]
Which test is crashing?

**Comment by slackner** [comment]
@c10ud: As already mentioned before your patch contained multiple issues. In v5 you have fixed one of them, but there is still the remaining issue that the resource memory is never reused, which is what happens most likely on Windows. In your patch you set heap_memory to NULL and managed to FALSE, so new memory will be allocated when it is needed again. Depending on the game this could lead to about twice the amount of memory usage. @titan73 is currently investigating how to solve this problem, based on my attempt in https://github.com/wine-compholio/wine-staging/issues/290#issuecomment-70757381.

**Comment by c10ud** [comment]
v6 - tests passing, games working...please read/test: https://github.com/c10ud/wine/commit/87164a18e36005527101e89f2af1561ec8142ba2

**Comment by titan73** [comment]
I was trying to excluse WINED3DUSAGE_RENDERTARGET and WINED3DUSAGE_DEPTHSTENCIL from managed texture and actually found out that usage could also be 0 which is problematic.
So indeed only managing texture that have the WINED3DUSAGE_TEXTURE is a better solution.
Why do you you need to add WINED3DUSAGE_OFFSCREEN in the wined3d interface since you check the TEXTURE flag?
You don't need 'FIXME("Tried to allocate RAM again in a managed resource\n");' as it is normal behaviour for managed mode.
Is the test crash fixed and is the accounting working correctly now?

**Comment by c10ud** [comment]
After excluding, as you said, WINED3DUSAGE_RENDERTARGET and WINED3DUSAGE_DEPTHSTENCIL I added the WINED3DUSAGE_OFFSCREEN flag since offscreen surfaces identify themselves as textures and had their heap set to null.
Instead of checking for a weird "0" usage, I find it nice to know what's going on.

as for the fixme, I guess future people would want, given this memory model, to improve their resource usage and avoid messing with ram they are guaranteed it's already there (?)
anyway not a biggie for me, i can leave it out

everything I tried looks good, if you can, test again please

**Comment by titan73** [comment]
Finally we can keep the fixme but display only once so the console will not be flooded.
That would be good to use a better message also, something like:
"Sub-resource managed by texture cannot be evicted. Global texture eviction to optimize system memory use not implemented yet"

**Comment by slackner** [comment]
>After excluding, as you said, WINED3DUSAGE_RENDERTARGET and >WINED3DUSAGE_DEPTHSTENCIL I added the WINED3DUSAGE_OFFSCREEN flag since >offscreen surfaces identify themselves as textures and had their heap set to null.

As you're never explicitly checking explicit for WINED3DUSAGE_OFFSCREEN i don't think its necessary to add this constant. Just leaving it zero might be easier, especially when taking into account that it still has to be combined with CSMT and other patches later.

I think it also wouldn't hurt to add a similar check for `if (desc->usage & WINED3DUSAGE_TEXTURE)` to all other (cubetexture, volumetexture) functions, just to be sure. Even if this part is probably never triggered with current wine versions, it might change in the future.

The FIXME will probably occur very often, either we should use something different here (fixme-once, trace, ...) or completely remove it.

One last final step: All the major issues should be fixed, but I am still concerned about users having trouble with memory usage. Would you mind adding a registry switch for that? It probably should be enabled by default, but if users experience problems (too much system memory allocated or weird crashes, ...) they can still disable it in the registry. With those changes included I think we can proceed with the CSMT case or alternatively, if this turns out to be too complicated, we could also just disable this patchset for CSMT for now (requires no changes to the patch itself). What do you guys think?

**Comment by c10ud** [comment]
Well for starters even if I'm not explicitly checking for it, creating a texture which is an offscreen surface with "anonymous" usage feels weird, hence the flag.
Also it doesn't "break" anything, and imho helps keeping things under control. 

As for adding the usage check for volumes and cubetextures, I'm against to that too since it could mask other subtle issues, just like the offscreen usage = 0 did. I'd say we modify what it needs to be modified, and other stuff will eventually trigger bugs, if any.

FIXME -- not occurring in my limited testing on CSMT lol and wow, probably will occur with CSMT? anyway, FIXME-ONCE would be good imo

As for memory usage: my tests are not scientific but I didn't notice anything weird. 
I wouldn't make this method selectable as it would only introduce further ifdefs and potential breakages.

Finally, CSMT. I still need to test it after the latest changes since I grew tired of switching branches back and forth (my old AMD isn't pleased). I'd say once we agree on the CSMT-less patch we move on and see what needs to be done for CSMT. I hope/think only setting the ->map_heap_memory on init is required, but ofcourse I may be wrong.

Also if anyone could get some review from the d3d bosses, that'd be great. I tried but no luck.

greets

**Comment by titan73** [comment]
I would expect the TEXTURE flag to be always set for cube and volume texture but that would probably be better to do the check for them too. Moreover if you change the check "if (desc->usage & WINED3DUSAGE_TEXTURE)" by "if (current_heap)" when computing surfaces heap, you can easily disable the feature from wined3d_texture_init. We will have them to check the registry only in wined3d_texture_init and resource_init.

**Comment by titan73** [comment]
Regarding global sysmem texture eviction, I think this can be achieved by using a counter set to the sub-resource count when initializing the texture and a callback that will be called whenever a sub-resource is freed or reallocated. This callback will be decreased for free and increased for alloc. When decremented to 0, we free the texture memory and reset all sub-resource heap and when incremented from 0, we reallocated the memory and recompute all sub-resources heap.

**Comment by slackner** [comment]
Short status update: @titan73 provided me with an updated patch including the improvements mentioned above. This means that the non-CSMT part should be fine (except the memory usage, which is not optimized yet). With the patches below the fix is disabled by default, and has to be enabled in the "Staging" tab. As a next step we should focus which changes are necessary for CSMT.

Latest version of the patches:

1. http://ix.io/fWE - implements contiguous mipmap support, but is disabled by default
2. https://raw.githubusercontent.com/wine-compholio/wine-staging/804b80bb9466abc7b3d8020412be5d4d54a3a4f6/patches/winecfg-Staging/0001-winecfg-Add-staging-tab-for-CSMT.patch
3. http://ix.io/fWF - Adds a checkbox to the winecfg tab.
(please note that the CSMT option doesn't work without additional patches, so keep the checkbox disabled)

I was thinking about adding these changes already in 1.7.35 for the non-CSMT part, but then discovered several problems, which have to be fixed first.

* @c10ud: I noticed that your patch doesn't include your real name. We're using the same [rules as for upstream wine](http://wiki.winehq.org/SubmittingPatches) here, so you'll have to fix that before I can add the patch. 

>GitWine has more details, but the key point is to use your real name. This isn't just for developers to show a stronger commitment and establish trust, but also helps discourage irresponsible programmers from submitting unacceptable code (such as dissassembled Microsoft components).

* I had a longer discussion with @DarkPlayer today and its not really trivial to enable patches just for "non-CSMT". Besides the fact that it might confuse the user when specific options cannot be combined, there is a high risk of breaking things when using "#ifdef" in multiple patches on the same file. For example nested "#ifdef"'s might have the effect that the logic doesn't work as desired, and might break in future versions without anyone noticing it. We decided that its too risky to do the necessary changes so shortly before the next release. I'll take a look at that during the next week so that we can theoretically enable it for the next one, but if we're able to write a proper fix for CSMT that would be even better. I quickly looked at it but I'll need some more time to figure out how we can ensure that this patchset still works correctly with CSMT.

Marking as milestone for 1.7.26 (with or without CSMT support).

**Comment by titan73** [comment]
Here is a rough implementation http://ix.io/fWX of the code release/realloc of the contiguous texture memory.I I just tested Chernobyl and got the traces I expected and no crash happened.
This is a not finished version and is here for review only.

**Comment by c10ud** [comment]
aww I really wished you wouldn't go with the option. If we get it to work OK, clean and good there's no reason to keep the old behaviour I think.

 As for the name I'll take care of updating the patches once we're good to go.

Lastly I'll wait for titan to stabilize his patch before making further comments but the whole selectable allocation option to me only makes everything more complicated than it needs to be..

**Comment by titan73** [comment]
When the feature will be reliable, the option will be removed. Currently there are 2 known problems that needs to be addressed. The memory footprint and csmt.
I tested my patch with some games and I verified it with traces that it works. I would like to see the difference of memory footprint but afaik wine does not have such feature. I tested also with Bio Shock 2 but even w/o any patches it freezes after few seconds so I cannot compared anything.

Regarding my patch I don't think it's more complicated but that does not mean it has the right design. I tried to do the minimal changes to make it work and keep it simple. I don't want to have a code invasive for a prototype. I think we need to fix the functional problems first and then discussed about the best design.

Regarding the design, if we want to have the correct one, we need to move all the code at texture level with some hooks (maybe texture_ops) in surface and volume. We should also get rid of the code in resource_init that determines if the resource is managed or not. This is the texture that should pass it to lower layers.
You may ask some hints to Stefan about what could be the best design.

Even if the memory footprint problem is not solved yet. The csmt problem should be investigated too.

**Comment by c10ud** [comment]
v7, includes my latest ideas about the alloc/free thing: https://github.com/c10ud/wine/commit/0a53d8ce38b7f592332afe7c98c572804d9b425a

**Comment by c10ud** [comment]
v8: https://github.com/c10ud/wine/commit/15108e236139b4190668563f8ba1e1491821719b

**Comment by c10ud** [comment]
[10:40] **stefand** c10ud: fwiw, I don't think any attempt that special-cases one resource type is going to work well
[10:40] **stefand** at least not for the upstream wine tree
[10:41] **stefand** container-managed allocation needs to be done for all resources and all pools
[10:41] **stefand** though I think without GL_ARB_BUFFER_STORAGE we need to allocate one GL PBO per subresource. But otherwise one allocation per container
[10:41] **stefand** and buffers would be their own container
[10:42] **stefand** that applies to all resources, regardless of type and pool
[10:42] **stefand** (and even the one PBO per subresource type allocation should be managed by the container, not the subresource)
[11:13] **c10ud** btw stefand (sorry) with the latest version of the patch I was special casing textures for surfaces because of offscreen surfaces (I added it as an usage type since 0 was ugly)...other than that buffers are the only remaining users afaik
[11:17] **c10ud** other than that, do you want to get entirely rid of the wined3d_resource struct for surfaces buffers and volumes?
[11:23] **stefand** pretty much, that is the idea
[11:23] **stefand** and add a wined3d_subresource instead
[11:23] **stefand** note that this is not that easy because currently ddraw, d3d8 and d3d9 depend on surface_from_resource and resource_from_surface and so on
[11:24] **stefand** ideally surfaces and volumes should go away from the public interface
[11:24] **stefand** but I don't think you have to go that far for your case
[11:24] **stefand** for transitioning purposes it might be ok to just add a wined3d_subresource, and make surfaces and volumes both resources and subresources at the same time
[11:25] **stefand** but don't make subresources a subclass of resources
[11:25] **stefand** (stringfellow said no-no to the subresource as a subclass of resource idea a year ago)

**Comment by c10ud** [comment]
Hi,
I tried following stefand's hints and started digging deeper in the code: it's like a ride in hell :)

However, you can find my current progress in this branch (latest 10 commits or so) https://github.com/c10ud/wine/commits/TextureAllocation
The commits aren't meaningful since I discovered the ugly things I had to do while I was changing the code (I should probably squash everything into 1 single commit..), however current state is that it sort of works (YMMV).
At the moment I'm battling with d3d9 tests, it's crashing somewhere around test_swapchain(), and I suspect, well, swapchains (use-after-free or something like that)

As usual comments, ideas, etc. are welcome

Ciao,
Riccardo

**Comment by c10ud** [comment]
Hi,

https://github.com/c10ud/wine/compare/TextureAllocation (warning: big diff)

This branch implements the minimal required modifications for resources to be allocated all at once in textures.
A new structure, wined3d_subresource is introduced for surfaces etc., and it does not manage its memory.
d3d* code has been adjusted.
Some codepaths are still hackish or flaky but overall it kind of works (needs some advice/look from more experienced d3d developers).

tested with: d3d8 tests, d3d9 tests, league of legends (d3d9 mode)

test it, break it and let me know!

**Comment by c10ud** [comment]
OK, I got some review and the response is... this patch is too big!

Now, since I'm unable to do further work on this, here's some hints from Stefan for breaking it into smaller patches:

* keep *both* struct wined3d_resource resource and struct wined3d_subresource subresource in struct wined3d_surface for transitioning purposes

and/or

*  remove surfaces and from the public API first and fix things like the reset loop before attempting to introduce a subresource structure

other than that the current code lies here: https://github.com/c10ud/wine/tree/TextureAllocation (full diff against a month old vanilla wine: https://github.com/wine-mirror/wine/compare/master...c10ud:TextureAllocation ).

If anyone's willing to do this (remember this change is kind of a stopper for CSMT) and needs some advice/has doubts about anything related to the code I modified, feel free to ask.

Ciao,
Riccardo

**Comment by slackner** [comment]
Hi everyone,

About one year ago, we announced that we would like to move to more advanced bug tracker. We are sorry for the inconvenience, but now its time to say finally goodbye to the old one. This means, all remaining opened bugs (including this one!) will be closed, and no longer available. If you still need any information from the old bugtracker, please write a mail to webmaster@fds-team.de.

More important, if your issue still exists, please open a new bug report at one of the places mentioned below. For bug reports, please attach terminal output from a recent Wine version (preferably >= 1.8). Please note that in the meantime, Wine Staging has become an official part of WineHQ, so if your issue is already tracked at winehq.org, there is NO NEED to open a new bug report for it. Feel free to add the Wine Staging maintainers (sebastian@fds-team.de, michael@fds-team.de) as CC if you want us to take a look at your bug report.

**NEW BUG TRACKERS:**

* For all Wine AND Wine Staging bugs. USE THIS IF IN DOUBT. (*)
  https://bugs.winehq.org/enter_bug.cgi?product=Wine

* For packaging issues:
  https://bugs.winehq.org/enter_bug.cgi?product=Packaging

* For patch submission:
  https://bugs.wine-staging.com/enter_bug.cgi

Best regards,
Sebastian

----

(*) The "Wine Staging" product in the winehq bugtracker should only be used, when your problem turns out to be a regression introduced in "Wine Staging". Since most problems also occur with Wine, it is recommended to open your bug in the "Wine" product, unless someone can show that it doesn't occur with the development branch.
